rust-src/ed25519_hd_key_derivation/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+use hmac::{Hmac, Mac};
+use regex::Regex;
+use sha2::Sha512;
+
+const ED25519_CURVE: &[u8; 12] = b"ed25519 seed";
+const HARDENED_OFFSET: u32 = 0x80000000;
+
+pub struct HdKeys {
+    pub key:    [u8; 32],
+    chain_code: [u8; 32],
+}
+
+fn get_master_key_from_seed(seed: &[u8]) -> HdKeys {
+    let mut mac = Hmac::<Sha512>::new_from_slice(ED25519_CURVE).unwrap();
+    mac.update(&seed);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn ckd_priv(keys: HdKeys, index: u32) -> HdKeys {
+    if index & HARDENED_OFFSET == 0 {
+        panic!("Received a non-hardened index!");
+    }
+
+    let mut data = vec![0u8];
+    data.extend_from_slice(&keys.key);
+    data.extend_from_slice(&index.to_be_bytes());
+
+    let mut mac = Hmac::<Sha512>::new_from_slice(&keys.chain_code).unwrap();
+    mac.update(&data);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn is_valid_path(path: &str) -> bool {
+    let path_regex = Regex::new("^m(/[0-9]+')+$").unwrap();
+    if !path_regex.is_match(path) {
+        return false;
+    }
+
+    let segments: Vec<&str> = path.split('/').collect();
+    segments
+        .iter()
+        .skip(1)
+        .map(|seg| seg.replace("'", ""))
+        .all(|seg| seg.parse::<u32>().is_ok())
+}
+
+/// Derives hierarchical deterministic keys for ed25519 according to the SLIP0010 (https://github.com/satoshilabs/slips/blob/master/slip-0010.md)
+/// specification.
+pub fn derive(path: &str, seed: &[u8]) -> Result<HdKeys, String> {
```

Comment:
It works better if errors are structured. Some time ago we started using `thiserror` for this. In this case it would look like
```rust
use thiserror::Error;
#[derive(Debug, Error)]
pub enum DeriveError {
    #[error("Invalid derivation path.")]
    InvalidPath,
    #[error("Seed must be between 128 and 512 bits.")]
    InvalidSeed,
}
```

Reply:
That's way better. Good to know.